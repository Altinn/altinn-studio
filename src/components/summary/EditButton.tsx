import React from 'react';

import { Button, ButtonColor, ButtonVariant } from '@digdir/design-system-react';
import { Edit } from '@navikt/ds-icons';

export interface IEditButtonProps {
  onClick: () => void;
  editText: string | null;
}

export function EditButton(props: IEditButtonProps) {
  return (
    <Button
      variant={ButtonVariant.Quiet}
      color={ButtonColor.Secondary}
      icon={<Edit aria-hidden={true} />}
      iconPlacement='right'
      onClick={props.onClick}
    >
      {props.editText}
    </Button>
  );
}
