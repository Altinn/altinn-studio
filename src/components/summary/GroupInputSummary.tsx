import * as React from 'react';

import { makeStyles, Typography } from '@material-ui/core';

import { useAppSelector } from 'src/common/hooks';
import { useDisplayData } from 'src/components/hooks';
import { useExpressions } from 'src/features/expressions/useExpressions';
import { getVariableTextKeysForRepeatingGroupComponent } from 'src/utils/formLayout';
import { getLanguageFromKey } from 'src/utils/sharedUtils';
import { getTextFromAppOrDefault } from 'src/utils/textResource';
import type { ITextResource, ITextResourceBindings } from 'src/types';

export interface ISingleInputSummary {
  index: number;
  formData: any;
  textResourceBindings: ITextResourceBindings | undefined;
  textResources: ITextResource[];
}

const useStyles = makeStyles({
  data: {
    fontWeight: 500,
    '& p': {
      fontWeight: 500,
    },
  },
  emptyField: {
    fontStyle: 'italic',
    fontSize: '1.4rem',
  },
});

function GroupInputSummary({ index, formData, textResourceBindings, textResources }: ISingleInputSummary) {
  const displayData = useDisplayData({ formData });
  const classes = useStyles();
  const language = useAppSelector((state) => state.language.language);

  const textResourceBindingsResolvedExpressions = useExpressions(textResourceBindings);

  const textResourceBindingsResolvedTextKeys = getVariableTextKeysForRepeatingGroupComponent(
    textResources,
    textResourceBindingsResolvedExpressions,
    index,
  );

  return (
    <Typography variant='body1'>
      <span>
        {textResourceBindingsResolvedTextKeys &&
          getTextFromAppOrDefault(textResourceBindingsResolvedTextKeys.title, textResources, {}, [], false)}
        {' : '}
      </span>
      {typeof displayData !== 'undefined' ? (
        <span className={classes.data}>{displayData}</span>
      ) : (
        <span className={classes.emptyField}>{getLanguageFromKey('general.empty_summary', language || {})}</span>
      )}
    </Typography>
  );
}

export default GroupInputSummary;
