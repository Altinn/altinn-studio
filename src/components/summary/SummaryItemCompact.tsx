import React from 'react';

import { makeStyles, Typography } from '@material-ui/core';

import { useAppSelector } from 'src/common/hooks/useAppSelector';
import { getLanguageFromKey } from 'src/language/sharedLanguage';
import { getTextFromAppOrDefault } from 'src/utils/textResource';
import type { LayoutNode } from 'src/utils/layout/hierarchy';

export interface ICompactSummaryItem {
  targetNode: LayoutNode;
  displayData: string;
}

const useStyles = makeStyles({
  data: {
    fontWeight: 500,
    '& p': {
      fontWeight: 500,
    },
  },
  emptyField: {
    fontStyle: 'italic',
    fontSize: '0.875rem',
  },
});

export function SummaryItemCompact({ targetNode, displayData }: ICompactSummaryItem) {
  const classes = useStyles();
  const language = useAppSelector((state) => state.language.language);
  const textResources = useAppSelector((state) => state.textResources.resources);
  const textBindings = targetNode.item.textResourceBindings;

  return (
    <Typography
      variant='body1'
      data-testid={'summary-item-compact'}
    >
      <span>
        {textBindings && getTextFromAppOrDefault(textBindings.title, textResources, {}, [], false)}
        {' : '}
      </span>
      {displayData ? (
        <span className={classes.data}>{displayData}</span>
      ) : (
        <span className={classes.emptyField}>{getLanguageFromKey('general.empty_summary', language || {})}</span>
      )}
    </Typography>
  );
}
