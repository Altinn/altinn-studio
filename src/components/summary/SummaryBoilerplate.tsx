import React from 'react';

import { Grid, makeStyles, Typography } from '@material-ui/core';
import cn from 'classnames';

import { EditButton } from 'src/components/summary/EditButton';
import { AltinnAppTheme } from 'src/theme/altinnAppTheme';
import type { ILayoutCompSummary } from 'src/layout/Summary/types';

export interface SummaryBoilerplateProps extends Omit<ILayoutCompSummary, 'type' | 'id'> {
  hasValidationMessages?: boolean;
  onChangeClick: () => void;
  changeText: string | null;
  label: any;
  readOnlyComponent?: boolean;
}

const useStyles = makeStyles({
  label: {
    fontWeight: 500,
    fontSize: '1.125rem',
    '& p': {
      fontWeight: 500,
      fontSize: '1.125rem',
    },
  },
  labelWithError: {
    color: AltinnAppTheme.altinnPalette.primary.red,
    '& p': {
      color: AltinnAppTheme.altinnPalette.primary.red,
    },
  },
});
export function SummaryBoilerplate({
  hasValidationMessages,
  onChangeClick,
  changeText,
  label,
  readOnlyComponent,
  display,
}: SummaryBoilerplateProps) {
  const classes = useStyles();
  const shouldShowChangeButton = !readOnlyComponent && !display?.hideChangeButton;
  return (
    <>
      <Grid
        item={true}
        xs={10}
        {...(hasValidationMessages && {
          'data-testid': 'has-validation-message',
        })}
      >
        <Typography
          variant='body1'
          className={cn(hasValidationMessages && !display?.hideValidationMessages && classes.labelWithError)}
          component='span'
        >
          {label}
        </Typography>
      </Grid>
      <Grid
        item
        xs={2}
      >
        {shouldShowChangeButton && (
          <EditButton
            onClick={onChangeClick}
            editText={changeText}
          />
        )}
      </Grid>
    </>
  );
}
