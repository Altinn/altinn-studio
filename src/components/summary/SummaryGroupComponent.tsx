import React from 'react';

import { Grid, makeStyles, Typography } from '@material-ui/core';
import cn from 'classnames';

import { useAppSelector } from 'src/common/hooks/useAppSelector';
import { ErrorPaper } from 'src/components/message/ErrorPaper';
import { EditButton } from 'src/components/summary/EditButton';
import { GroupInputSummary } from 'src/components/summary/GroupInputSummary';
import { DisplayGroupContainer } from 'src/features/form/containers/DisplayGroupContainer';
import { renderLayoutComponent } from 'src/features/form/containers/Form';
import { getLanguageFromKey } from 'src/language/sharedLanguage';
import { ComponentType } from 'src/layout';
import { getLayoutComponentObject } from 'src/layout/LayoutComponent';
import { AltinnAppTheme } from 'src/theme/altinnAppTheme';
import { getDisplayFormDataForComponent, getFormDataForComponentInRepeatingGroup } from 'src/utils/formComponentUtils';
import { useResolvedNode } from 'src/utils/layout/ExprContext';
import { getTextFromAppOrDefault } from 'src/utils/textResource';
import type { ComponentFromSummary } from 'src/features/form/containers/DisplayGroupContainer';
import type { ILayoutGroup } from 'src/layout/Group/types';
import type { ComponentExceptGroupAndSummary, ILayoutComponent } from 'src/layout/layout';
import type { SummaryDisplayProperties } from 'src/layout/Summary/types';
import type { IRuntimeState } from 'src/types';
import type { AnyNode } from 'src/utils/layout/hierarchy.types';

export interface ISummaryGroupComponent {
  pageRef?: string;
  componentRef?: string;
  changeText: string | null;
  onChangeClick: () => void;
  largeGroup?: boolean;
  display?: SummaryDisplayProperties;
  excludedChildren?: string[];
}

const gridStyle = {
  paddingTop: '12px',
};

const useStyles = makeStyles({
  border: {
    border: '2px solid #EFEFEF',
    marginTop: 12,
    marginBottom: 12,
    padding: 12,
    '@media print': {
      pageBreakInside: 'avoid',
    },
  },
  label: {
    fontWeight: 500,
    fontSize: '1.125rem',
    '& p': {
      fontWeight: 500,
      fontSize: '1.125rem',
    },
  },
  labelWithError: {
    color: AltinnAppTheme.altinnPalette.primary.red,
    '& p': {
      color: AltinnAppTheme.altinnPalette.primary.red,
    },
  },
  link: {
    background: 'none',
    border: 'none',
    borderBottom: '2px solid #008FD6',
    cursor: 'pointer',
    paddingLeft: 0,
  },
  emptyField: {
    fontStyle: 'italic',
    fontSize: '1rem',
    marginTop: 4,
  },
});

export function SummaryGroupComponent({
  pageRef,
  componentRef,
  largeGroup,
  onChangeClick,
  changeText,
  display,
  excludedChildren,
}: ISummaryGroupComponent) {
  const classes = useStyles();

  const node = useResolvedNode(componentRef);
  const textResourceBindings = node?.item.textResourceBindings;

  const removeExcludedChildren = (n: AnyNode<'resolved'>) =>
    !excludedChildren ||
    !excludedChildren.includes(n.item.id) ||
    !excludedChildren.includes(`${n.item.baseComponentId}`);

  const repeatingGroups = useAppSelector((state) => state.formLayout.uiConfig.repeatingGroups);
  const formData = useAppSelector((state) => state.formData.formData);
  const textResources = useAppSelector((state) => state.textResources.resources);
  const language = useAppSelector((state) => state.language.language);
  const options = useAppSelector((state) => state.optionState.options);
  const attachments = useAppSelector((state: IRuntimeState) => state.attachments.attachments);
  const validations = useAppSelector((state) => state.formValidations.validations);
  const hiddenFields = useAppSelector((state) => new Set(state.formLayout.uiConfig.hiddenFields));

  const title = React.useMemo(() => {
    if (textResources && textResourceBindings) {
      const titleKey = textResourceBindings.title;
      if (!titleKey) {
        return '';
      }

      return getTextFromAppOrDefault(titleKey, textResources, language || {}, [], true);
    }

    return '';
  }, [textResources, textResourceBindings, language]);

  const groupHasErrors = React.useMemo(() => {
    if (!largeGroup && node && pageRef) {
      return node.flat(true).some((child) =>
        Object.keys(validations[pageRef]?.[child.item.id] || {}).some((bindingKey: string) => {
          const length = validations[pageRef][child.item.id][bindingKey]?.errors?.length;
          return length && length > 0;
        }),
      );
    }

    return false;
  }, [node, largeGroup, pageRef, validations]);

  const createRepeatingGroupSummaryComponents = () => {
    if (!node || !('rows' in node.item)) {
      return [];
    }

    const componentArray: JSX.Element[] = [];
    for (const row of node.item.rows) {
      if (!row) {
        continue;
      }
      const childSummaryComponents = node
        .children(undefined, row.index)
        .filter(removeExcludedChildren)
        .filter(
          (node) =>
            getLayoutComponentObject(node.item.type as ComponentExceptGroupAndSummary)?.getComponentType() ===
            ComponentType.Form,
        )
        .map((n) => {
          if (n.isHidden(hiddenFields)) {
            return;
          }

          const formDataForComponent = getDisplayFormDataForComponent(
            formData,
            attachments,
            n.item as ILayoutComponent,
            textResources,
            options,
            repeatingGroups,
          );

          return (
            <GroupInputSummary
              key={n.item.id}
              componentId={n.item.id}
              formData={formDataForComponent}
              textResources={textResources}
            />
          );
        });
      componentArray.push(
        <div
          key={row.index}
          className={classes.border}
        >
          {childSummaryComponents}
        </div>,
      );
    }

    return componentArray;
  };

  const createRepeatingGroupSummaryForLargeGroups = () => {
    if (!node || !('rows' in node.item)) {
      return;
    }

    const componentArray: JSX.Element[] = [];
    for (const row of node.item.rows) {
      if (!row) {
        continue;
      }

      const groupContainer = {
        ...node.item,
        children: [],
      } as ILayoutGroup;

      const childSummaryComponents: ComponentFromSummary[] = [];
      node
        .children(undefined, row.index)
        .filter(removeExcludedChildren)
        .forEach((n) => {
          if (n.isHidden(hiddenFields)) {
            return;
          }

          const summaryId = `${n.item.id}-summary${n.item.type === 'Group' ? '-group' : ''}`;
          const groupDataModelBinding = node?.item.dataModelBindings?.group?.replace(/\[\d+]/g, '');
          let formDataForComponent: any;
          if (n.item.type !== 'Group') {
            formDataForComponent = getFormDataForComponentInRepeatingGroup(
              formData,
              attachments,
              n.item as ILayoutComponent,
              row.index,
              groupDataModelBinding,
              textResources,
              options,
              repeatingGroups,
            );
          }
          groupContainer.children.push(summaryId);

          const summaryComponent: ComponentFromSummary = {
            id: summaryId,
            type: 'Summary',
            componentRef: n.item.id,
            pageRef: pageRef,
            dataModelBindings: {},
            textResourceBindings: {},
            readOnly: false,
            required: false,
            formData: formDataForComponent,
            display,
            excludedChildren,
          };

          childSummaryComponents.push(summaryComponent);
        });

      componentArray.push(
        <DisplayGroupContainer
          key={`${groupContainer.id}-summary`}
          id={`${groupContainer.id}-${row.index}-summary`}
          components={childSummaryComponents}
          container={groupContainer}
          renderLayoutComponent={renderLayoutComponent}
        />,
      );
    }
    return componentArray;
  };

  const renderComponents = largeGroup
    ? createRepeatingGroupSummaryForLargeGroups()
    : createRepeatingGroupSummaryComponents();

  if (!language || !renderComponents) {
    return null;
  }

  const isEmpty = renderComponents?.length <= 0;

  if (!isEmpty && largeGroup) {
    // Tricking our return type to be JSX.Element. Apparently, returning an array causes problems elsewhere.
    return renderComponents as unknown as JSX.Element;
  }

  return (
    <>
      <Grid
        container={true}
        data-testid={'summary-group-component'}
      >
        <Grid
          item={true}
          xs={10}
        >
          <Typography
            variant='body1'
            className={cn(classes.label, groupHasErrors && !display?.hideValidationMessages && classes.labelWithError)}
            component='span'
          >
            {title}
          </Typography>
        </Grid>
        <Grid
          item
          xs={2}
        >
          {!display?.hideChangeButton && (
            <EditButton
              onClick={onChangeClick}
              editText={changeText}
            />
          )}
        </Grid>
        <Grid
          item
          xs={12}
        >
          {isEmpty ? (
            <Typography
              variant='body1'
              className={classes.emptyField}
              component='p'
            >
              {getLanguageFromKey('general.empty_summary', language)}
            </Typography>
          ) : (
            renderComponents
          )}
        </Grid>
      </Grid>
      {groupHasErrors && !display?.hideValidationMessages && (
        <Grid
          container={true}
          style={gridStyle}
        >
          <ErrorPaper message={getLanguageFromKey('group.row_error', language)} />
          <Grid
            item={true}
            xs={12}
          >
            {!display?.hideChangeButton && (
              <button
                className={classes.link}
                onClick={onChangeClick}
                type='button'
              >
                {getTextFromAppOrDefault('form_filler.summary_go_to_correct_page', textResources, language, [], true)}
              </button>
            )}
          </Grid>
        </Grid>
      )}
    </>
  );
}
