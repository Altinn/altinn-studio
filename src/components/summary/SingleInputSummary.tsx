import React from 'react';

import { Grid, makeStyles, Typography } from '@material-ui/core';

import { useAppSelector } from 'src/common/hooks/useAppSelector';
import { useDisplayData } from 'src/components/hooks/useDisplayData';
import { SummaryBoilerplate } from 'src/components/summary/SummaryBoilerplate';
import { getLanguageFromKey } from 'src/language/sharedLanguage';
import type { SummaryDisplayProperties } from 'src/layout/Summary/types';

export interface ISingleInputSummary {
  formData: any;
  label: any;
  hasValidationMessages: boolean;
  changeText: any;
  onChangeClick: () => void;
  readOnlyComponent?: boolean;
  display?: SummaryDisplayProperties;
}

const useStyles = makeStyles({
  data: {
    fontWeight: 500,
    fontSize: '1.125rem',
    '& p': {
      fontWeight: 500,
      fontSize: '1.125rem',
    },
  },
  emptyField: {
    fontStyle: 'italic',
    fontSize: '1rem',
  },
});

export function SingleInputSummary({ formData, display, ...rest }: ISingleInputSummary) {
  const classes = useStyles();
  const displayData = useDisplayData({ formData });
  const language = useAppSelector((state) => state.language.language);

  return (
    <>
      <SummaryBoilerplate
        display={display}
        {...rest}
      />
      <Grid
        item
        xs={12}
        data-testid={'single-input-summary'}
      >
        {typeof displayData !== 'undefined' ? (
          <Typography
            className={classes.data}
            variant='body1'
          >
            {displayData}
          </Typography>
        ) : (
          <Typography
            variant='body1'
            className={classes.emptyField}
          >
            {getLanguageFromKey('general.empty_summary', language || {})}
          </Typography>
        )}
      </Grid>
    </>
  );
}
