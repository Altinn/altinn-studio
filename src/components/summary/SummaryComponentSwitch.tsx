import React from 'react';

import { SingleInputSummary } from 'src/components/summary/SingleInputSummary';
import { SummaryBoilerplate } from 'src/components/summary/SummaryBoilerplate';
import { SummaryGroupComponent } from 'src/components/summary/SummaryGroupComponent';
import { MultipleChoiceSummary } from 'src/layout/Checkboxes/MultipleChoiceSummary';
import { AttachmentSummaryComponent } from 'src/layout/FileUpload/AttachmentSummaryComponent';
import { AttachmentWithTagSummaryComponent } from 'src/layout/FileUploadWithTag/AttachmentWithTagSummaryComponent';
import { MapComponentSummary } from 'src/layout/Map/MapComponentSummary';
import { useResolvedNode } from 'src/utils/layout/ExprContext';
import type { ILayoutGroup } from 'src/layout/Group/types';
import type { ILayoutComponent } from 'src/layout/layout';
import type { ILayoutCompSummary } from 'src/layout/Summary/types';

export interface ISummaryComponentSwitch extends Omit<ILayoutCompSummary, 'type'> {
  change: {
    onChangeClick: () => void;
    changeText: string | null;
  };
  formComponent?: ILayoutComponent | ILayoutGroup;
  hasValidationMessages?: boolean;
  label?: JSX.Element | JSX.Element[] | null | undefined;
  formData?: any;
  groupProps?: {
    pageRef?: string;
    largeGroup?: boolean;
  };
}

export function SummaryComponentSwitch({
  change,
  formComponent,
  label,
  componentRef,
  hasValidationMessages,
  formData,
  groupProps = {},
  display,
}: ISummaryComponentSwitch) {
  const resolved = useResolvedNode(formComponent)?.item;

  if (!formComponent) {
    return null;
  }

  const hasDataBindings = Object.keys(formComponent.dataModelBindings || {}).length === 0;

  if (hasDataBindings && formComponent.type === 'FileUpload' && componentRef) {
    return (
      <>
        <SummaryBoilerplate
          {...change}
          label={label}
          hasValidationMessages={hasValidationMessages}
          display={display}
        />
        <AttachmentSummaryComponent componentRef={componentRef} />
      </>
    );
  }

  if (hasDataBindings && formComponent.type === 'FileUploadWithTag' && componentRef) {
    return (
      <>
        <SummaryBoilerplate
          {...change}
          label={label}
          hasValidationMessages={hasValidationMessages}
          display={display}
        />
        <AttachmentWithTagSummaryComponent
          componentRef={componentRef}
          component={formComponent}
        />
      </>
    );
  }

  if (formComponent.type === 'Group') {
    return (
      <SummaryGroupComponent
        {...change}
        {...groupProps}
        componentRef={componentRef}
        display={display}
      />
    );
  }

  if (formComponent.type === 'Checkboxes' && typeof formData !== 'string') {
    return (
      <MultipleChoiceSummary
        {...change}
        label={label}
        hasValidationMessages={!!hasValidationMessages}
        formData={formData}
        readOnlyComponent={resolved?.readOnly}
        display={display}
      />
    );
  }

  if (formComponent.type === 'Map') {
    return (
      <>
        <SummaryBoilerplate
          {...change}
          label={label}
          hasValidationMessages={!!hasValidationMessages}
          display={display}
        />
        <MapComponentSummary
          component={formComponent}
          formData={formData}
        />
      </>
    );
  }

  return (
    <SingleInputSummary
      {...change}
      label={label}
      hasValidationMessages={!!hasValidationMessages}
      formData={formData}
      readOnlyComponent={resolved?.readOnly}
      display={display}
    />
  );
}
