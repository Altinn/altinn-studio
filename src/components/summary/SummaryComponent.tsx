import * as React from 'react';
import { shallowEqual } from 'react-redux';

import { Grid, makeStyles } from '@material-ui/core';
import cn from 'classnames';

import { useAppDispatch, useAppSelector } from 'src/common/hooks';
import ErrorPaper from 'src/components/message/ErrorPaper';
import SummaryComponentSwitch from 'src/components/summary/SummaryComponentSwitch';
import { useExpressionsForComponent } from 'src/features/expressions/useExpressions';
import { DisplayGroupContainer } from 'src/features/form/containers/DisplayGroupContainer';
import { mapGroupComponents } from 'src/features/form/containers/formUtils';
import { FormLayoutActions } from 'src/features/form/layout/formLayoutSlice';
import { makeGetHidden } from 'src/selectors/getLayoutData';
import printStyles from 'src/styles/print.module.css';
import {
  componentHasValidationMessages,
  getComponentValidations,
  getDisplayFormDataForComponent,
} from 'src/utils/formComponentUtils';
import { getTextFromAppOrDefault } from 'src/utils/textResource';
import type { ILayoutComponent } from 'src/layout/layout';
import type { ILayoutCompSummary } from 'src/layout/Summary/types';
import type { IComponentValidations, IRuntimeState } from 'src/types';

export interface ISummaryComponent extends Omit<ILayoutCompSummary, 'type'> {
  parentGroup?: string;
  index?: number;
  formData?: any;
}

const useStyles = makeStyles({
  border: {
    marginBottom: 10,
    paddingBottom: 10,
    borderBottom: '1px dashed #008FD6',
  },
  link: {
    background: 'none',
    border: 'none',
    borderBottom: '2px solid #008FD6',
    cursor: 'pointer',
    paddingLeft: 0,
  },
});

export function SummaryComponent(_props: ISummaryComponent) {
  const { id, grid, ...summaryProps } = _props;
  const { componentRef, display, ...groupProps } = summaryProps;
  const { index, pageRef, formData, ...containerProps } = _props;
  const container = { ...containerProps, type: 'Summary' } as ILayoutCompSummary;
  const classes = useStyles();
  const dispatch = useAppDispatch();
  const GetHiddenSelector = makeGetHidden();
  const [componentValidations, setComponentValidations] = React.useState<IComponentValidations>({});
  const [hasValidationMessages, setHasValidationMessages] = React.useState(false);
  const hidden = useAppSelector((state) => {
    if (GetHiddenSelector(state, { id })) {
      return true;
    }
    return !!(componentRef && GetHiddenSelector(state, { id: componentRef }));
  });
  const summaryPageName = useAppSelector((state) => state.formLayout.uiConfig.currentView);
  const changeText = useAppSelector(
    (state) =>
      state.language.language &&
      getTextFromAppOrDefault(
        'form_filler.summary_item_change',
        state.textResources.resources,
        state.language.language,
        [],
        true,
      ),
  );
  const formValidations = useAppSelector((state) => state.formValidations.validations);
  const layout = useAppSelector((state) =>
    state.formLayout.layouts && pageRef ? state.formLayout.layouts[pageRef] : undefined,
  );
  const attachments = useAppSelector((state: IRuntimeState) => state.attachments.attachments);
  const _formComponent = useAppSelector((state) => {
    return (
      (state.formLayout.layouts && pageRef && state.formLayout.layouts[pageRef]?.find((c) => c.id === componentRef)) ||
      undefined
    );
  });
  const formComponent = useExpressionsForComponent(_formComponent);
  const summaryComponent = useExpressionsForComponent(container);

  const goToCorrectPageLinkText = useAppSelector((state) => {
    return (
      state.language.language &&
      getTextFromAppOrDefault(
        'form_filler.summary_go_to_correct_page',
        state.textResources.resources,
        state.language.language,
        [],
        true,
      )
    );
  });
  const calculatedFormData = useAppSelector((state) => {
    if (formComponent?.type === 'Group') {
      return undefined;
    }
    if (
      (formComponent?.type === 'FileUpload' || formComponent?.type === 'FileUploadWithTag') &&
      Object.keys(formComponent.dataModelBindings || {}).length === 0
    ) {
      return undefined;
    }
    return (
      formData ||
      getDisplayFormDataForComponent(
        state.formData.formData,
        attachments,
        formComponent as ILayoutComponent,
        state.textResources.resources,
        state.optionState.options,
        state.formLayout.uiConfig.repeatingGroups,
        true,
      )
    );
  }, shallowEqual);
  const label = useAppSelector((state) => {
    const titleKey = formComponent?.textResourceBindings?.title;
    if (titleKey) {
      return (
        state.language.language &&
        getTextFromAppOrDefault(titleKey, state.textResources.resources, state.language.language, [], false)
      );
    }
    return undefined;
  });

  const onChangeClick = () => {
    if (!pageRef) {
      return;
    }

    dispatch(
      FormLayoutActions.updateCurrentView({
        newView: pageRef,
        returnToView: summaryPageName,
        focusComponentId: componentRef,
      }),
    );
  };

  React.useEffect(() => {
    if (formComponent && formComponent.type !== 'Group') {
      const componentId = typeof index === 'number' && index >= 0 ? `${componentRef}-${index}` : componentRef;
      const validations =
        (componentId && pageRef && getComponentValidations(formValidations, componentId, pageRef)) || undefined;
      setComponentValidations(validations || {});
      setHasValidationMessages(componentHasValidationMessages(validations));
    }
  }, [formValidations, layout, pageRef, formComponent, componentRef, index]);

  if (hidden) {
    return null;
  }
  const change = {
    onChangeClick,
    changeText,
  };

  const isNonRepeatingGroup =
    formComponent?.type === 'Group' && (!formComponent.maxCount || formComponent.maxCount <= 1);
  if (isNonRepeatingGroup) {
    // Display children as summary components
    const groupComponents = mapGroupComponents(formComponent, layout);
    return (
      <DisplayGroupContainer
        key={id}
        container={formComponent}
        components={groupComponents}
        renderLayoutComponent={(child) => (
          <SummaryComponent
            id={`__summary__${child.id}`}
            componentRef={child.id}
            pageRef={groupProps.pageRef}
            largeGroup={groupProps.largeGroup}
            display={display}
          />
        )}
      />
    );
  }

  const displayGrid = display && display.useComponentGrid ? formComponent?.grid : grid;
  return (
    <Grid
      item={true}
      xs={displayGrid?.xs || 12}
      sm={displayGrid?.sm || false}
      md={displayGrid?.md || false}
      lg={displayGrid?.lg || false}
      xl={displayGrid?.xl || false}
      data-testid={`summary-${id}`}
      className={cn({
        [printStyles['break-before']]: summaryComponent.pageBreak?.breakBefore,
        [printStyles['break-after']]: summaryComponent.pageBreak?.breakAfter,
      })}
    >
      <Grid
        container={true}
        className={cn({
          [classes.border]: !display?.hideBottomBorder,
        })}
      >
        <SummaryComponentSwitch
          id={id}
          change={change}
          formComponent={formComponent}
          label={label}
          hasValidationMessages={hasValidationMessages}
          formData={calculatedFormData}
          componentRef={componentRef}
          groupProps={groupProps}
          display={display}
        />
        {hasValidationMessages && !display?.hideValidationMessages && (
          <Grid
            container={true}
            style={{ paddingTop: '12px' }}
            spacing={2}
          >
            {Object.keys(componentValidations).map((binding: string) =>
              componentValidations[binding]?.errors?.map((validationText: string) => (
                <ErrorPaper
                  key={`key-${validationText}`}
                  message={validationText}
                />
              )),
            )}
            <Grid
              item={true}
              xs={12}
            >
              {!display?.hideChangeButton && (
                <button
                  className={classes.link}
                  onClick={onChangeClick}
                  type='button'
                >
                  {goToCorrectPageLinkText}
                </button>
              )}
            </Grid>
          </Grid>
        )}
      </Grid>
    </Grid>
  );
}
