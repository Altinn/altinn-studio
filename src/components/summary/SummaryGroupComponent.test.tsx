import React from 'react';

import configureStore from 'redux-mock-store';

import { getFormDataStateMock } from 'src/__mocks__/formDataStateMock';
import { getFormLayoutStateMock } from 'src/__mocks__/formLayoutStateMock';
import { getInitialStateMock } from 'src/__mocks__/initialStateMock';
import { SummaryGroupComponent } from 'src/components/summary/SummaryGroupComponent';
import { renderWithProviders } from 'src/testUtils';
import type { ISummaryGroupComponent } from 'src/components/summary/SummaryGroupComponent';

describe('SummaryGroupComponent', () => {
  let mockHandleDataChange: () => void;
  let mockStore;

  beforeAll(() => {
    const createStore = configureStore();
    const formLayout = getFormLayoutStateMock({
      layouts: {
        page1: [
          {
            type: 'Group',
            id: 'groupComponent',
            dataModelBindings: {
              group: 'mockGroup',
            },
            textResourceBindings: {
              title: 'mockGroupTitle',
            },
            children: ['0:mockId1', '1:mockId2'],
            edit: {
              multiPage: true,
            },
            maxCount: 3,
          },
          {
            type: 'Input',
            id: 'mockId1',
            dataModelBindings: {
              simpleBinding: 'mockGroup.mockDataBinding1',
            },
            readOnly: false,
            required: false,
            disabled: false,
            textResourceBindings: {
              title: 'mockField1',
            },
            triggers: [],
          },
          {
            type: 'Input',
            id: 'mockId2',
            dataModelBindings: {
              simpleBinding: 'mockGroup.mockDataBinding2',
            },
            readOnly: false,
            required: false,
            disabled: false,
            textResourceBindings: {
              title: 'mockField2',
            },
            triggers: [],
          },
        ],
      },
      uiConfig: {
        autoSave: true,
        focus: null,
        hiddenFields: [],
        repeatingGroups: {
          groupComponent: {
            index: 0,
            dataModelBinding: 'mockGroup',
          },
        },
        currentView: 'FormLayout',
        navigationConfig: {},
        tracks: {
          order: [],
          hidden: [],
          hiddenExpr: {},
        },
        excludePageFromPdf: [],
        excludeComponentFromPdf: [],
      },
    });

    const formData = getFormDataStateMock({
      formData: {
        'mockGroup[0].mockDataBinding1': '1',
        'mockGroup[0].mockDataBinding2': '2',
      },
    });

    const initialState: any = getInitialStateMock({
      formData,
      formLayout,
      textResources: {
        error: null,
        language: 'nb',
        resources: [
          {
            id: 'mockGroupTitle',
            value: 'Mock group',
          },
          {
            id: 'mockField1',
            value: 'Mock field 1',
          },
          {
            id: 'mockField2',
            value: 'Mock field 2',
          },
        ],
      },
    });
    mockStore = createStore(initialState);
  });

  beforeEach(() => {
    mockHandleDataChange = jest.fn();
  });

  test('SummaryGroupComponent -- should match snapshot', () => {
    const { asFragment } = renderSummaryGroupComponent();
    expect(asFragment()).toMatchSnapshot();
  });

  function renderSummaryGroupComponent(props: Partial<ISummaryGroupComponent> = {}) {
    const defaultProps: ISummaryGroupComponent = {
      pageRef: 'page1',
      componentRef: 'groupComponent',
      largeGroup: false,
      changeText: 'Change',
      onChangeClick: mockHandleDataChange,
    };

    return renderWithProviders(
      <SummaryGroupComponent
        {...defaultProps}
        {...props}
      />,
      { store: mockStore },
    );
  }
});
