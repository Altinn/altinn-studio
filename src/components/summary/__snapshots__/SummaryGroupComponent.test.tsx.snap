// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SummaryGroupComponent SummaryGroupComponent -- should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container"
    data-testid="summary-group-component"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10"
    >
      <span
        class="MuiTypography-root makeStyles-label-2 MuiTypography-body1"
      >
        Mock group
      </span>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
    >
      <button
        class="Button-module_button__2ZuB7 Button-module_small__l39oh Button-module_quiet__1KlhF Button-module_secondary__R0waJ"
        type="button"
      >
        Change
        <svg
          aria-hidden="true"
          class="Button-module_icon__-43u5"
          fill="none"
          focusable="false"
          height="1em"
          role="img"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M22.835 1.165a3.976 3.976 0 0 1 0 5.623L8.073 21.549.682 24 0 23.318l2.45-7.392L17.21 1.165a3.977 3.977 0 0 1 5.624 0Zm-4.218 7.029-2.811-2.812L4.188 17l-1.393 4.205 4.207-1.395L18.618 8.194ZM21.43 2.57a1.989 1.989 0 0 0-2.703-.1l-.108.1-1.406 1.406 2.811 2.812 1.406-1.406a1.988 1.988 0 0 0 .101-2.703l-.1-.109Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="makeStyles-border-1"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          <span>
            <span>
              Mock field 1
            </span>
             : 
          </span>
          <span
            class="makeStyles-data-6"
          >
            1
          </span>
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          <span>
            <span>
              Mock field 2
            </span>
             : 
          </span>
          <span
            class="makeStyles-data-6"
          >
            2
          </span>
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
